src/porepy/grids/grid.py

```diff
@@ -422,83 +423,94 @@ def nrm(u):
+        self.face_centers = 0.5 * self.nodes * np.abs(fn_orient)
+
+        # Compute the temporary cell centers as average of the cell nodes
+        faceno, cellno, cf_orient = sps.find(self.cell_faces)
+        cx = np.bincount(cellno, weights=self.face_centers[0, faceno])
+        cy = np.bincount(cellno, weights=self.face_centers[1, faceno])
+        cz = np.bincount(cellno, weights=self.face_centers[2, faceno])
+        temp_cell_centers = np.vstack((cx, cy, cz)) / np.bincount(cellno)
+
+        # Create sub-simplexes based on (face, cell center) pairs.
+        # Compute the vectors that is normal to the sub-simplex and whose length is the
+        # area.
+        subsimplex_heights = self.face_centers[:, faceno] - temp_cell_centers[:, cellno]
+        subsimplex_normals = 0.5 * np.cross(
+            subsimplex_heights, cf_orient * tangent[:, faceno], axis=0
         )
-        self.face_centers = 0.5 * (xe1 + xe2)
-        n = edge_length_z.shape[0]
-        self.face_normals = np.vstack((edge_length_y, -edge_length_x, np.zeros(n)))
-
-        cell_faces, cellno = self.cell_faces.nonzero()
-        cx = np.bincount(cellno, weights=self.face_centers[0, cell_faces])
-        cy = np.bincount(cellno, weights=self.face_centers[1, cell_faces])
-        cz = np.bincount(cellno, weights=self.face_centers[2, cell_faces])
-        self.cell_centers = np.vstack((cx, cy, cz)) / np.bincount(cellno)
-
-        a = xe1[:, cell_faces] - self.cell_centers[:, cellno]
-        b = xe2[:, cell_faces] - self.cell_centers[:, cellno]
-
-        sub_volumes = 0.5 * np.abs(a[0] * b[1] - a[1] * b[0])
-        self.cell_volumes = np.bincount(cellno, weights=sub_volumes).astype(float)
 
+        # Construct the unit normal of the grid as planar object
+        if is_oriented:
+            plane_normal = subsimplex_normals.sum(axis=1)
+            plane_normal /= np.linalg.norm(plane_normal)
+        else:
+            plane_normal = pp.map_geometry.compute_normal(self.nodes)
+
+        # Compute the face normals by rotating the tangent according to the orientation
+        # of the plane
+        self.face_normals = np.cross(tangent, plane_normal, axis=0)
+
+        # Compute the signed volumes of sub-simplexes.
+        # If the nodes are counter clock-wise these are positive values.
+        subsimplex_volumes = np.dot(plane_normal, subsimplex_normals)
+
+        # In case of inconsistent orientation,
+        # the sub-simplex volumes and normals need to be corrected
+        if not is_oriented:
+            subsimplex_volumes = np.abs(subsimplex_volumes)
+
+            # We flip the normal if the inner product between
+            # the height (face_center - cell_center)
+            # and the face normal is negative
+            flip = (
+                cf_orient
+                * np.sum(subsimplex_heights * self.face_normals[:, faceno], axis=0)
+            ) < 0
+            flip = np.bincount(faceno, weights=flip).astype(bool)
```

Comment:
I don't see how you make sure that you only do this from one side (+ or -) for internal faces?

Reply:
The bincount takes care of this. At first, we compute flip for every face-cell pair as a boolean. These are added in a bincount to construct flip of length num_faces. The astype(bool) ensures that duplicates resulting from both sides (+ and -) are counted as a single True